# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1365491868 14400
#      Tue Apr 09 03:17:48 2013 -0400
# Node ID 98155e2b6d42a008ad1d2d678dda35ff6b9560fe
# Parent  0696dcc92fc8a7f54a8f6536e33288ee798182aa
get terminal size on window systems

* lo-sysdep.h, lo-sysdep.cc (w32_terminal_rows, w32_terminal_cols):
New functions.
* cmd-edit.cc (gnu_readline::do_terminal_rows,
gnu_readline::do_terminal_cols): Use them.

diff --git a/liboctave/system/lo-sysdep.cc b/liboctave/system/lo-sysdep.cc
--- a/liboctave/system/lo-sysdep.cc
+++ b/liboctave/system/lo-sysdep.cc
@@ -135,16 +135,46 @@ octave_popen2 (const std::string& cmd, c
   CloseHandle (childRead);
   CloseHandle (childWrite);
   CloseHandle (pi.hProcess);
   CloseHandle (pi.hThread);
 
   return pid;
 }
 
+int
+w32_terminal_rows (void)
+{
+  int retval = 0;
+
+  HANDLE console = GetStdHandle (STD_OUTPUT_HANDLE);
+
+  CONSOLE_SCREEN_BUFFER_INFO csbi; 
+
+  if (GetConsoleScreenBufferInfo (console, &csbi))
+    retval = csbi.dwSize.Y;
+
+  return retval;
+}
+
+int
+w32_terminal_cols (void)
+{
+  int retval = 0;
+
+  HANDLE console = GetStdHandle (STD_OUTPUT_HANDLE);
+
+  CONSOLE_SCREEN_BUFFER_INFO csbi; 
+
+  if (GetConsoleScreenBufferInfo (console, &csbi))
+    retval = csbi.dwSize.X;
+
+  return retval;
+}
+
 void
 w32_clear_console_window (void)
 {
   HANDLE console = GetStdHandle (STD_OUTPUT_HANDLE);
 
   // Get the number of character cells in the current buffer.
 
   CONSOLE_SCREEN_BUFFER_INFO csbi; 
diff --git a/liboctave/system/lo-sysdep.h b/liboctave/system/lo-sysdep.h
--- a/liboctave/system/lo-sysdep.h
+++ b/liboctave/system/lo-sysdep.h
@@ -31,13 +31,15 @@ class string_vector;
 extern std::string octave_getcwd (void);
 
 extern int octave_chdir (const std::string&);
 
 #if defined (__WIN32__) && ! defined (__CYGWIN__)
 extern pid_t octave_popen2 (const std::string&, const string_vector&,
     bool, int *, std::string&);
 
+extern int w32_terminal_rows (void);
+extern int w32_terminal_cols (void);
 extern void w32_clear_console_window (void);
 
 #endif
 
 #endif
diff --git a/liboctave/util/cmd-edit.cc b/liboctave/util/cmd-edit.cc
--- a/liboctave/util/cmd-edit.cc
+++ b/liboctave/util/cmd-edit.cc
@@ -297,25 +297,33 @@ gnu_readline::do_get_output_stream (void
 // of being correct even if the window changes size (they may be
 // wrong if, for example, the luser changes the window size while the
 // pager is running, and the signal is handled by the pager instead of
 // us.
 
 int
 gnu_readline::do_terminal_rows (void)
 {
+#if defined (__WIN32__) && ! defined (__CYGWIN__)
+  int sh = w32_terminal_rows ();
+#else
   int sh = ::octave_rl_screen_height ();
+#endif
 
   return sh > 0 ? sh : 24;
 }
 
 int
 gnu_readline::do_terminal_cols (void)
 {
+#if defined (__WIN32__) && ! defined (__CYGWIN__)
+  int sw = w32_terminal_cols ();
+#else
   int sw = ::octave_rl_screen_width ();
+#endif
 
   return sw > 0 ? sw : 80;
 }
 
 void
 gnu_readline::do_clear_screen (void)
 {
 #if defined (__WIN32__) && ! defined (__CYGWIN__)
